mesonbuild/ast/printer.py

```diff
@@ -0,0 +1,203 @@
+# Copyright 2019 The Meson development team
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This class contains the basic functionality needed to run any interpreter
+# or an interpreter-based tool
+
+from .. import mparser
+from . import AstVisitor
+import re
+
+arithmic_map = {
+    'add': '+',
+    'sub': '-',
+    'mod': '%',
+    'mul': '*',
+    'div': '/'
+}
+
+class AstPrinter(AstVisitor):
+    def __init__(self, indent: int = 2, arg_newline_cutoff: int = 5):
+        self.result = ''
+        self.indent = indent
+        self.arg_newline_cutoff = arg_newline_cutoff
+        self.ci = ''
+        self.is_newline = True
+        self.last_level = 0
+
+    def post_process(self):
+        self.result = re.sub(r'\s+\n', '\n', self.result)
+
+    def append(self, data: str, node: mparser.BaseNode):
+        level = 0
+        if node and hasattr(node, 'level'):
+            level = node.level
+        else:
+            level = self.last_level
+        self.last_level = level
+        if self.is_newline:
+            self.result += ' ' * (level * self.indent)
+        self.result += data
+        self.is_newline = False
+
+    def appendS(self, data: str, node: mparser.BaseNode):
```

Comment:
This method could have a clearer name. It's not obvious what `appendS` stands for.

Reply:
Renamed the function to `append_padded`.